### DISCUSSION TITLE
PBPotentiometer Invert

### DISCUSSION THREAD

**Discussion by TsarNikolai2** [tttapa/Control-Surface#820] [Q&A] [answered]
Hi,

I see the PBPotentiometer has the option of inverting the output, I wired mine backwards and it works, but I'd like to invert the output without changing the wiring and I'm not sure how to get it to work. I don't know much about coding, but I did write this from examples and everything else is working.

```cpp
#include <Control_Surface.h> // Include the Control Surface library
#include <MIDI_Outputs/CCButtonLatching.hpp>
#define LED 14  // The pin the LED is connected to

// Instantiate a MIDI over USB interface.
USBMIDI_Interface midi;

// Instantiate a PBPotentiometer object
PBPotentiometer potentiometer {
  A6,        // Analog pin connected to potentiometer
  CHANNEL_1, // MIDI Channel 1
  };

CCPotentiometer pots[]  { 
{A0, 7}, 
{A1, 10},
{A2, 112},
{A3, 113},
{A7, 114},
{A8, 115},
//{A9, 116},
{A10, 117},

};

// Instantiate a CCButton object
CCButtonLatching button { 
//  Push button on pin 5: 
  2,
//   General Purpose Controller #1 on MIDI channel 1:
  {MIDI_CC::General_Purpose_Controller_1, CHANNEL_1},
};

void setup() {
  Control_Surface.begin(); // Initialize Control Surface
  pinMode(LED, OUTPUT); // Declare the LED as an output
}

void loop() {
  Control_Surface.loop(); // Update the Control Surface
  digitalWrite(LED, LOW); // Turn the LED on
}
```

**Comment by tttapa** [comment, marked as answer]
You can add the line `potentiometer.invert();` to the beginning of your `setup()` function.

[PBPotentiometer::invert()](https://tttapa.github.io/Control-Surface-doc/Doxygen/da/d6a/classPBPotentiometer.html#a7fa1616cc61c19a5efcc863c950f7f30)

**Reply by TsarNikolai2** [reply]
Thank you, that worked perfectly! Perhaps a good note to add to the PBPotentiometer Example page. Thank you for this library too, it's been easy to work with even for a hack like me ;)

**Comment by benwadub** [comment]
hello, is it possible to have this function for an encoder to revert it?

**Reply by tttapa** [reply]
To change the direction of an encoder you simply swap the A and B pins.

**Reply by benwadub** [reply]
yes that what I thought at first, but if there was a possibility to do it via software this could be faster as I have 16 encoders on the prototype on two mcp23017 :-)

**Reply by tttapa** [reply]
You can swap the pins in software as well, just change the order of the constructor arguments.

**Comment by benwadub** [comment]
I can t access it with the library using mcp23017, or I m missing something?

**Reply by tttapa** [reply]
In that case, just set the speed multiplier to -1 instead of 1. See https://tttapa.github.io/Control-Surface/Doxygen/d7/dac/MCP23017-RotaryEncoder-No-Interrupts_8ino-example.html.

**Reply by benwadub** [reply]
Thanks a lot!!

**Comment by onlyv8** [comment]
Hi,

To begin, let's me thank Tttapa for his great job. I've built a real control surface with no difficulties. This control surface is use with[ Synth1 VST](https://daichilab.sakura.ne.jp/softsynth/index.html):

![IMG_20250607_190406](https://github.com/user-attachments/assets/e4d0f20a-ac34-4bca-94da-f3ffeb0f6cc4)

It work very fine, with a little exception: all my potentiometers is inverted and analog read value too...

And modify these problems is not easy:

![IMG_20250615_170130](https://github.com/user-attachments/assets/456269e5-a9bd-457b-85f5-9f29e849c3ca)

All my potentiometer is welded with the copper tape...

Where and how i must correct the code ?

i take two days to find the answers and the syntax on the wonderful Control surface documentations. But without clear explaination...

But i'm a newbie, sorry

I've just read 30 minutes ago the "potentiometer.invert();" fonction, but i do not know how and where to code it.

Here my code:

`
#include <Control_Surface.h> // Include the Control Surface library


// Select the serial port to use.
auto &serial = Serial;
// Instantiate a Serial MIDI interface at the default MIDI baud rate.
SerialMIDI_Interface<decltype(serial)> midi {serial, MIDI_BAUD};

// Instantiate an analog multiplexer
CD74HC4067 mux1 { A1,{3, 4, 5, 6} };


CCPotentiometer Potentiometers1[] {
  { mux1.pin(0), { 15 } },
  { mux1.pin(1), { 20 } },
  { mux1.pin(2), { 21 } },
  { mux1.pin(3), { 28 } },
  { mux1.pin(4), { 29 } },
  { mux1.pin(5), { 31 } },
  { mux1.pin(6), { 35 } },
  { mux1.pin(7), { 47 } },
  { mux1.pin(8), { 48 } },
  { mux1.pin(9), { 49 } },
  { mux1.pin(10), { 50 } },
  { mux1.pin(11), { 52 } },
  { mux1.pin(12), { 53 } },
  { mux1.pin(13), { 54 } },
  { mux1.pin(14), { 55 } },
  { mux1.pin(15), { 56 } },
};

CD74HC4067 mux2 {
  A2, {7, 8, 9, 10}
};
 

CCPotentiometer Potentiometers2[] {
  { mux2.pin(0), { 57 } },
  { mux2.pin(1), { 58 } },
  { mux2.pin(2), { 59 } },
  { mux2.pin(3), { 60 } },
  { mux2.pin(4), { 62 } },
  { mux2.pin(5), { 63 } },
  { mux2.pin(6), { 70 } },
  { mux2.pin(7), { 71 } },
  { mux2.pin(8), { 72 } },
  { mux2.pin(9), { 73 } },
  { mux2.pin(10), { 75 } },
  { mux2.pin(11), { 79 } },
  { mux2.pin(12), { 84 } },
  { mux2.pin(13), { 85 } },
  { mux2.pin(14), { 86 } },
  { mux2.pin(15), { 90 } },
};

CD74HC4067 mux3 {
  A3, {11, 12, 13, 14} 
};


CCButton button [] {
  { mux3.pin(0), { 23 } },
  { mux3.pin(1), { 25 } },
  { mux3.pin(2), { 26 } },
  { mux3.pin(3), { 27 } },
  { mux3.pin(4), { 30 } },
  { mux3.pin(5), { 61 } },
  { mux3.pin(6), { 92 } },
  { mux3.pin(7), { 103 } },
  { mux3.pin(8), { 104 } },
  { mux3.pin(9), { 105 } },
  { mux3.pin(10), { 110 } },
  { mux3.pin(11), { 111 } },
  { mux3.pin(12), { 87 } },
  { mux3.pin(13), { 76 } },
  { mux3.pin(14), { 77 } },
};

CCAbsoluteEncoder  enc [] {
  {{15, 16}, { 14 } },
  {{17, 18}, { 22 } },
  {{19, 20}, { 24 } },
  {{21, 22}, { 41 } },
  {{24, 25}, { 46 } },
  {{26, 27}, { 51 } },
  {{28, 29}, { 93 } },
  {{30, 31}, { 94 } },
  {{32, 33}, { 106 } },
  {{34, 35}, { 107 } },
  {{36, 37}, { 88 } },
  {{38, 39}, { 89 } },
  {{40, 41}, { 74 } },
  {{42, 43}, { 78 } },

};

CCPotentiometer pot [] {

  { A4, { 91 } },
  { A5, { 95 } },
  { A6, { 102 } },
  { A7, { 108 } },
  { A8, { 109 } },
  { A9, { 118 } },
};
 
void setup() {
  Control_Surface.begin();  // Initialize the Control Surface
}
 
void loop() {
  Control_Surface.loop();  // Update the Control Surface
}`

With all my thanks for the help.

**Reply by onlyv8** [reply]
Just find the solution:


#include <Control_Surface.h> // Include the Control Surface library


// Select the serial port to use.
auto &serial = Serial;
// Instantiate a Serial MIDI interface at the default MIDI baud rate.
SerialMIDI_Interface<decltype(serial)> midi {serial, MIDI_BAUD};

// Instantiate an analog multiplexer
CD74HC4067 mux1 { A1,{3, 4, 5, 6} };


CCPotentiometer Potentiometers1[] {
  { mux1.pin(0), { 15 } },
  { mux1.pin(1), { 20 } },
  { mux1.pin(2), { 21 } },
  { mux1.pin(3), { 28 } },
  { mux1.pin(4), { 29 } },
  { mux1.pin(5), { 31 } },
  { mux1.pin(6), { 35 } },
  { mux1.pin(7), { 47 } },
  { mux1.pin(8), { 48 } },
  { mux1.pin(9), { 49 } },
  { mux1.pin(10), { 50 } },
  { mux1.pin(11), { 52 } },
  { mux1.pin(12), { 53 } },
  { mux1.pin(13), { 54 } },
  { mux1.pin(14), { 55 } },
  { mux1.pin(15), { 56 } },
};

CD74HC4067 mux2 {
  A2, {7, 8, 9, 10}
};
 

CCPotentiometer Potentiometers2[] {
  { mux2.pin(0), { 57 } },
  { mux2.pin(1), { 58 } },
  { mux2.pin(2), { 59 } },
  { mux2.pin(3), { 60 } },
  { mux2.pin(4), { 62 } },
  { mux2.pin(5), { 63 } },
  { mux2.pin(6), { 70 } },
  { mux2.pin(7), { 71 } },
  { mux2.pin(8), { 72 } },
  { mux2.pin(9), { 73 } },
  { mux2.pin(10), { 75 } },
  { mux2.pin(11), { 79 } },
  { mux2.pin(12), { 84 } },
  { mux2.pin(13), { 85 } },
  { mux2.pin(14), { 86 } },
  { mux2.pin(15), { 90 } },
};

CD74HC4067 mux3 {
  A3, {11, 12, 13, 14} 
};


CCButton button [] {
  { mux3.pin(0), { 23 } },
  { mux3.pin(1), { 25 } },
  { mux3.pin(2), { 26 } },
  { mux3.pin(3), { 27 } },
  { mux3.pin(4), { 30 } },
  { mux3.pin(5), { 61 } },
  { mux3.pin(6), { 92 } },
  { mux3.pin(7), { 103 } },
  { mux3.pin(8), { 104 } },
  { mux3.pin(9), { 105 } },
  { mux3.pin(10), { 110 } },
  { mux3.pin(11), { 111 } },
  { mux3.pin(12), { 87 } },
  { mux3.pin(13), { 76 } },
  { mux3.pin(14), { 77 } },
};

CCAbsoluteEncoder  enc [] {
  {{15, 16}, { 14 } },
  {{17, 18}, { 22 } },
  {{19, 20}, { 24 } },
  {{21, 22}, { 41 } },
  {{24, 25}, { 46 } },
  {{26, 27}, { 51 } },
  {{28, 29}, { 93 } },
  {{30, 31}, { 94 } },
  {{32, 33}, { 106 } },
  {{34, 35}, { 107 } },
  {{36, 37}, { 88 } },
  {{38, 39}, { 89 } },
  {{40, 41}, { 74 } },
  {{42, 43}, { 78 } },

};

CCPotentiometer pot [] {

  { A4, { 91 } },
  { A5, { 95 } },
  { A6, { 102 } },
  { A7, { 108 } },
  { A8, { 109 } },
  { A9, { 118 } },
};
 
void setup() {
  Control_Surface.begin();  // Initialize the Control Surface
 for (auto &potentiometer : Potentiometers1)
    potentiometer.invert();

 for (auto &potentiometer : Potentiometers2)
    potentiometer.invert();   

 for (auto &potentiometer : pot)
    potentiometer.invert(); 
  
}
 
void loop() {
  Control_Surface.loop();  // Update the Control Surface
}
